app/scripts/metamask-controller.js

```diff
@@ -2422,6 +2446,19 @@ export default class MetamaskController extends EventEmitter {
 
         for (const [origin, accounts] of changedAccounts.entries()) {
           this._notifyAccountsChange(origin, accounts);
+          // When permissions are added for a given origin we need to
+          // set the network client id for that origin to the current
+          // globally selected network client id.
+          if (accounts.length > 0) {
+            // TODO add this logic directly in the SelectedNetworkController
+            this.selectedNetworkController.setNetworkClientIdForDomain(
+              origin,
+              this.networkController.state.selectedNetworkClientId,
+            );
+          } else {
+            // TODO remove domain from selectedNetworkController when permissions are removed
+            // Add this logic directly in the SelectedNetworkController
+          }
```

Comment:
There is a WIP PR against the `SelectedNetworkController` to pull this logic into that controller instead: https://github.com/MetaMask/core/pull/3969